Sources/TuistAutomation/Device/DeviceController.swift

```diff
@@ -0,0 +1,168 @@
+import Command
+import FileSystem
+import Foundation
+import Mockable
+import Path
+import TuistSupport
+import XcodeGraph
+
+enum DeviceControllerError: FatalError {
+    case applicationVerificationFailed
+
+    var description: String {
+        switch self {
+        case .applicationVerificationFailed:
+            "The app could not be installed because the verification failed. Make sure that your device is registered in your Apple Developer account."
+        }
+    }
+
+    var type: ErrorType {
+        switch self {
+        case .applicationVerificationFailed:
+            .abort
+        }
+    }
+}
+
+@Mockable
+/// Utility to interact with the `devicectl` CLI.
+public protocol DeviceControlling {
+    func findAvailableDevices() async throws -> [PhysicalDevice]
+    func installApp(
+        at path: AbsolutePath,
+        device: PhysicalDevice
+    ) async throws
+    func launchApp(
+        bundleId: String,
+        device: PhysicalDevice
+    ) async throws
+}
+
+public final class DeviceController: DeviceControlling {
+    private let fileSystem: FileSysteming
+    private let commandRunner: CommandRunning
+
+    public init(
+        fileSystem: FileSysteming = FileSystem(),
+        commandRunner: CommandRunning = CommandRunner()
+    ) {
+        self.fileSystem = fileSystem
+        self.commandRunner = commandRunner
+    }
+
+    public func findAvailableDevices() async throws -> [PhysicalDevice] {
+        try await fileSystem.runInTemporaryDirectory(prefix: "device-controller-find-available-devices") { temporaryPath in
+            let devicesListOutputPath = temporaryPath.appending(component: "devices-list-output-path.json")
+            _ = try await commandRunner.run(
+                arguments: [
+                    "/usr/bin/xcrun", "devicectl",
+                    "list", "devices",
+                    "--json-output", devicesListOutputPath.pathString,
```

Comment:
It'd be awesome if they supported sending the JSON through stdout.

Reply:
it would ... but they don't 😢